In [83]:
import numpy as np
import pandas as pd

In [271]:
horse_data = pd.read_csv('horse_data.csv', na_values='?')

In [272]:
horse_data.head()

2  1   530101  38.50     66    28    3  3.1    ?  2.1  ...  45.00  8.40  \
0  1.0  1   534817   39.2   88.0  20.0  NaN  NaN  4.0  1.0  ...   50.0  85.0   
1  2.0  1   530334   38.3   40.0  24.0  1.0  1.0  3.0  1.0  ...   33.0   6.7   
2  1.0  9  5290409   39.1  164.0  84.0  4.0  1.0  6.0  2.0  ...   48.0   7.2   
3  2.0  1   530255   37.3  104.0  35.0  NaN  NaN  6.0  2.0  ...   74.0   7.4   
4  2.0  1   528355    NaN    NaN   NaN  2.0  1.0  3.0  1.0  ...    NaN   NaN   

   ?.4  ?.5  2.2  2.3  11300  00000  00000.1  2.4  
0  2.0  2.0  3.0    2   2208      0        0    2  
1  NaN  NaN  1.0    2      0      0        0    1  
2  3.0  5.3  2.0    1   2208      0        0    1  
3  NaN  NaN  2.0    2   4300      0        0    2  
4  NaN  NaN  1.0    2      0      0        0    2  

[5 rows x 28 columns]

In [273]:
horse_data.columns = [
  'surgery',
  'age',
  'hospital_number',
  'rectal_temp',
  'pulse',
  'respiratory_rate',
  'temp_of_extremities',
  'peripheral_pulse',
  'mucous_membranes',
  'capillary_refill_time',
  'pain',
  'peristalsis',
  'abdominal_distension',
  'nasogastric_tube',
  'nasogastric_reflux',
  'nasogastric_reflux_ph',
  'rectal_examination_feces',
  'abdomen',
  'packed_cell_volume',
  'total_protein',
  'abdominocentesis_appearance',
  'abdomcentesis_total_protein',
  'outcome',
  'surgical_lesion',
  'lesion_1',
  'lesion_2',
  'lesion_3',
  'cp_data'
]

In [274]:
horse_data = horse_data.astype({
  'surgery': 'category',
  'age': 'category',
  'hospital_number': 'category',
  'rectal_temp': 'float64',
  'pulse': 'float64',
  'respiratory_rate': 'float64',
  'temp_of_extremities': 'category',
  'peripheral_pulse': 'category',
  'mucous_membranes': 'category',
  'capillary_refill_time': 'category',
  'pain': 'category',
  'peristalsis': 'category',
  'abdominal_distension': 'category',
  'nasogastric_tube': 'category',
  'nasogastric_reflux': 'category',
  'nasogastric_reflux_ph': 'float64',
  'rectal_examination_feces': 'category',
  'abdomen': 'category',
  'packed_cell_volume': 'float64',
  'total_protein': 'float64',
  'abdominocentesis_appearance': 'category',
  'abdomcentesis_total_protein': 'float64',
  'outcome': 'category',
  'surgical_lesion': 'category',
  'lesion_1': 'category',
  'lesion_2': 'category',
  'lesion_3': 'category',
  'cp_data': 'category'
})

In [275]:
horse_data.head()

surgery age hospital_number  rectal_temp  pulse  respiratory_rate  \
0     1.0   1          534817         39.2   88.0              20.0   
1     2.0   1          530334         38.3   40.0              24.0   
2     1.0   9         5290409         39.1  164.0              84.0   
3     2.0   1          530255         37.3  104.0              35.0   
4     2.0   1          528355          NaN    NaN               NaN   

  temp_of_extremities peripheral_pulse mucous_membranes capillary_refill_time  \
0                 NaN              NaN              4.0                   1.0   
1                 1.0              1.0              3.0                   1.0   
2                 4.0              1.0              6.0                   2.0   
3                 NaN              NaN              6.0                   2.0   
4                 2.0              1.0              3.0                   1.0   

   ... packed_cell_volume total_protein abdominocentesis_appearance  \
0  ...               50.0          85.0                         2.0   
1  ...               33.0           6.7                         NaN   
2  ...               48.0           7.2                         3.0   
3  ...               74.0           7.4                         NaN   
4  ...                NaN           NaN                         NaN   

  abdomcentesis_total_protein outcome  surgical_lesion lesion_1 lesion_2  \
0                         2.0     3.0                2     2208        0   
1                         NaN     1.0                2        0        0   
2                         5.3     2.0                1     2208        0   
3                         NaN     2.0                2     4300        0   
4                         NaN     1.0                2        0        0   

   lesion_3  cp_data  
0         0        2  
1         0        1  
2         0        1  
3         0        2  
4         0        2  

[5 rows x 28 columns]

In [276]:
horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype   
---  ------                       --------------  -----   
 0   surgery                      298 non-null    category
 1   age                          299 non-null    category
 2   hospital_number              299 non-null    category
 3   rectal_temp                  239 non-null    float64 
 4   pulse                        275 non-null    float64 
 5   respiratory_rate             241 non-null    float64 
 6   temp_of_extremities          243 non-null    category
 7   peripheral_pulse             230 non-null    category
 8   mucous_membranes             253 non-null    category
 9   capillary_refill_time        267 non-null    category
 10  pain                         244 non-null    category
 11  peristalsis                  255 non-null    category
 12  abdominal_distension         243 non-null    category
 13  nasog

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

Выберем 8 стобцов для дальнейшего изучения

In [277]:
horse_short_data = horse_data[[
  'rectal_temp',
  'pulse',
  'respiratory_rate',
  'total_protein',
  'nasogastric_reflux_ph',
  'pain',
  'peripheral_pulse',
  'mucous_membranes',
]]

horse_short_data.head()

rectal_temp  pulse  respiratory_rate  total_protein  nasogastric_reflux_ph  \
0         39.2   88.0              20.0           85.0                    NaN   
1         38.3   40.0              24.0            6.7                    NaN   
2         39.1  164.0              84.0            7.2                    5.0   
3         37.3  104.0              35.0            7.4                    NaN   
4          NaN    NaN               NaN            NaN                    NaN   

  pain peripheral_pulse mucous_membranes  
0  3.0              NaN              4.0  
1  3.0              1.0              3.0  
2  2.0              1.0              6.0  
3  NaN              NaN              6.0  
4  2.0              1.0              3.0

In [278]:
horse_short_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype   
---  ------                 --------------  -----   
 0   rectal_temp            239 non-null    float64 
 1   pulse                  275 non-null    float64 
 2   respiratory_rate       241 non-null    float64 
 3   total_protein          266 non-null    float64 
 4   nasogastric_reflux_ph  53 non-null     float64 
 5   pain                   244 non-null    category
 6   peripheral_pulse       230 non-null    category
 7   mucous_membranes       253 non-null    category
dtypes: category(3), float64(5)
memory usage: 13.3 KB


In [10]:
horse_short_data.describe()

rectal_temp       pulse  respiratory_rate  total_protein  \
count   239.000000  275.000000        241.000000     266.000000   
mean     38.166527   71.934545         30.427386      24.517293   
std       0.733508   28.680522         17.678256      27.509056   
min      35.400000   30.000000          8.000000       3.300000   
25%      37.800000   48.000000         18.000000       6.500000   
50%      38.200000   64.000000         24.000000       7.500000   
75%      38.500000   88.000000         36.000000      57.000000   
max      40.800000  184.000000         96.000000      89.000000   

       nasogastric_reflux_ph  
count              53.000000  
mean                4.707547  
std                 1.982311  
min                 1.000000  
25%                 3.000000  
50%                 5.000000  
75%                 6.500000  
max                 7.500000

In [11]:
def get_metrics(series):
  column_type = series.dtype
  
  if column_type == 'float64':
    min = series.min()
    max = series.max()
    range = max - min
    mean = series.mean()
    median = series.median()
    std = series.std()
    var = series.var()
    mode = series.mode()
    q1 = series.quantile(0.25)
    q3 = series.quantile(0.75)
    iqr = q3 - q1
  
    return {
      'min': round(min, 2),
      'max': round(max, 2),
      'range': round(range, 2),
      'mean': round(mean, 2),
      'median': round(median, 2),
      'std': round(std, 2),
      'var': round(var, 2),
      'mode': list(mode),
      'q1': round(q1, 2),
      'q3': round(q3, 2),
      'iqr': round(iqr, 2),
    }
  
  if column_type == 'category':
    mode = series.mode()
    
    return {
      'mode': list(mode),
    }

#### Расчет базовых метрик для столбца Rectal temperature

In [12]:
get_metrics(horse_short_data.rectal_temp)

{'min': 35.4,
 'max': 40.8,
 'range': 5.4,
 'mean': 38.17,
 'median': 38.2,
 'std': 0.73,
 'var': 0.54,
 'mode': [38.0],
 'q1': 37.8,
 'q3': 38.5,
 'iqr': 0.7}

- Минимальное значение - **35.4**
- Максимальное значение - **40.8**
- Размах - **5.4**
- Среднее значение - **38.17**
- Медиана - **38.2**
- Стандартное отклонение - **0.73**
- Дисперсия - **0.54**
- Мода - **38**
- Первый квартиль - **37.8**
- Третий квартиль - **38.5**
- Межквартильный размах - **0.7**

#### Расчет базовых метрик для столбца Pulse

In [13]:
get_metrics(horse_short_data.pulse)

{'min': 30.0,
 'max': 184.0,
 'range': 154.0,
 'mean': 71.93,
 'median': 64.0,
 'std': 28.68,
 'var': 822.57,
 'mode': [48.0],
 'q1': 48.0,
 'q3': 88.0,
 'iqr': 40.0}

- Минимальное значение - **30**
- Максимальное значение - **184**
- Размах - **154**
- Среднее значение - **71.93**
- Медиана - **64**
- Стандартное отклонение - **26.68**
- Дисперсия - **822.57**
- Мода - **48**
- Первый квартиль - **48**
- Третий квартиль - **88**
- Межквартильный размах - **40**

#### Расчет базовых метрик для столбца Respiratory rate

In [14]:
get_metrics(horse_short_data.respiratory_rate)

{'min': 8.0,
 'max': 96.0,
 'range': 88.0,
 'mean': 30.43,
 'median': 24.0,
 'std': 17.68,
 'var': 312.52,
 'mode': [20.0],
 'q1': 18.0,
 'q3': 36.0,
 'iqr': 18.0}

- Минимальное значение - **8**
- Максимальное значение - **96**
- Размах - **88**
- Среднее значение - **30.43**
- Медиана - **24**
- Стандартное отклонение - **17.68**
- Дисперсия - **312.52**
- Мода - **20**
- Первый квартиль - **18**
- Третий квартиль - **36**
- Межквартильный размах - **18**

#### Расчет базовых метрик для столбца Total protein

In [15]:
get_metrics(horse_short_data.total_protein)

{'min': 3.3,
 'max': 89.0,
 'range': 85.7,
 'mean': 24.52,
 'median': 7.5,
 'std': 27.51,
 'var': 756.75,
 'mode': [6.5, 7.0],
 'q1': 6.5,
 'q3': 57.0,
 'iqr': 50.5}

- Минимальное значение - **3.3**
- Максимальное значение - **89**
- Размах - **85.7**
- Среднее значение - **24.52**
- Медиана - **7.5**
- Стандартное отклонение - **27.51**
- Дисперсия - **756.75**
- Моды - **6.5** и **7**
- Первый квартиль - **6.5**
- Третий квартиль - **57**
- Межквартильный размах - **50.5**

#### Расчет базовых метрик для столбца Nasogastric reflux PH

In [16]:
get_metrics(horse_short_data.nasogastric_reflux_ph)

{'min': 1.0,
 'max': 7.5,
 'range': 6.5,
 'mean': 4.71,
 'median': 5.0,
 'std': 1.98,
 'var': 3.93,
 'mode': [2.0],
 'q1': 3.0,
 'q3': 6.5,
 'iqr': 3.5}

- Минимальное значение - **1**
- Максимальное значение - **7.5**
- Размах - **6.5**
- Среднее значение - **4.71**
- Медиана - **5**
- Стандартное отклонение - **1.98**
- Дисперсия - **3.93**
- Мода - **2**
- Первый квартиль - **3**
- Третий квартиль - **6.5**
- Межквартильный размах - **3.5**

#### Расчет базовых метрик для столбца Pain

In [17]:
get_metrics(horse_short_data.pain)

{'mode': [3.0]}

- Мода - **3** (Прерывистая легкая боль)

#### Расчет базовых метрик для столбца Peripheral pulse

In [18]:
get_metrics(horse_short_data.peripheral_pulse)

{'mode': [1.0]}

- Мода - **1** (Нормальный)

#### Расчет базовых метрик для столбца Mucous membranes

In [19]:
get_metrics(horse_short_data.mucous_membranes)

{'mode': [1.0]}

- Мода - **1** (Нормальный розовый)

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [20]:
def get_outliers(series):
  q1 = series.quantile(0.25)
  q3 = series.quantile(0.75)
  iqr = q3 - q1
  
  lower_bound = q1 - 1.5 * iqr
  upper_bound = q3 + 1.5 * iqr
  
  without_outliers = series[series.between(lower_bound, upper_bound, inclusive=True)]

  outliers = pd.concat([series, without_outliers]).drop_duplicates(keep=False)
  
  return (without_outliers, outliers)

#### Выбросы в столбце Rectal temperature

In [21]:
rectal_temp_without_outliers, rectal_temp_outliers = get_outliers(horse_short_data.rectal_temp)
rectal_temp_outliers

19     39.9
43     35.4
74     39.7
79     36.4
98     39.6
140    36.0
237    36.1
250    36.6
258    40.8
280    40.0
Name: rectal_temp, dtype: float64

#### Изменение средних

In [22]:
print('Среднее', horse_short_data.rectal_temp.mean())
print('Среднее без выбросов', rectal_temp_without_outliers.mean())

Среднее 38.166527196652716
Среднее без выбросов 38.16755555555556


In [23]:
print('Медиана', horse_short_data.rectal_temp.median())
print('Медиана без выбросов', rectal_temp_without_outliers.median())

Медиана 38.2
Медиана без выбросов 38.2


##### В описании столбца указано, что повышенная температура может быть из-за инфекции, а пониженная может быть, когда животное находится в шоке. В данном случае все значения, определенные как выбросы, находятся в реальных пределах. Так что их оставляем

#### Выбросы в столбце Pulse

In [24]:
pulse_without_outliers, pulse_outliers = get_outliers(horse_short_data.pulse)
pulse_outliers

2      164.0
54     160.0
254    184.0
Name: pulse, dtype: float64

In [25]:
print('Среднее', horse_short_data.pulse.mean())
print('Среднее без выбросов', pulse_without_outliers.mean())

Среднее 71.93454545454546
Среднее без выбросов 70.27407407407408


In [26]:
print('Медиана', horse_short_data.pulse.median())
print('Медиана без выбросов', pulse_without_outliers.median())

Медиана 64.0
Медиана без выбросов 64.0


##### Из описания столбца следует, что нормальные показатели для взрослых особей колеблются в пределах 30-40 ударов в минуту. Однако также сказано, что у болеющих животных, а также животных пребывающих в шоке (возможно коротком), значения могут быть повышенными. Думаю, что несмотря на то, что приведенные выше значения значительно сильно выбиваются из допустимых значений, их можно оставить

#### Выбросы в столбце Respiratory rate

In [27]:
respiratory_rate_without_outliers, respiratory_rate_outliers = get_outliers(horse_short_data.respiratory_rate)
respiratory_rate_outliers

40     72.0
119    66.0
243    88.0
Name: respiratory_rate, dtype: float64

In [28]:
print('Среднее', horse_short_data.respiratory_rate.mean())
print('Среднее без выбросов', respiratory_rate_without_outliers.mean())

Среднее 30.42738589211618
Среднее без выбросов 26.709821428571427


In [29]:
print('Медиана', horse_short_data.respiratory_rate.median())
print('Медиана без выбросов', respiratory_rate_without_outliers.median())

Медиана 24.0
Медиана без выбросов 24.0


##### В описании столбца указано, что полезность этих данных сомнительна из-за больших колебаний. Думаю, что найденные выбросы как раз подходят под колебания, так что их можно оставить

#### Выбросы в стоблце Total protein

In [31]:
total_protein_without_outliers, total_protein_outliers = get_outliers(horse_short_data.total_protein)
total_protein_outliers

Series([], Name: total_protein, dtype: float64)

##### Здесь выбросов нет

### Выбросы в столбце Nasogastric reflux PH

In [32]:
nasogastric_reflux_ph_without_outliers, nasogastric_reflux_ph_outliers = get_outliers(horse_short_data.nasogastric_reflux_ph)
nasogastric_reflux_ph_outliers

Series([], Name: nasogastric_reflux_ph, dtype: float64)

##### Здесь выбросов нет

### Задание 3. Работа с пропусками

Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [279]:
horse_short_data_original = horse_short_data.copy()

In [281]:
horse_short_data_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype   
---  ------                 --------------  -----   
 0   rectal_temp            239 non-null    float64 
 1   pulse                  275 non-null    float64 
 2   respiratory_rate       241 non-null    float64 
 3   total_protein          266 non-null    float64 
 4   nasogastric_reflux_ph  53 non-null     float64 
 5   pain                   244 non-null    category
 6   peripheral_pulse       230 non-null    category
 7   mucous_membranes       253 non-null    category
dtypes: category(3), float64(5)
memory usage: 13.3 KB


---

#### Пропуски в столбце Rectal temperature

In [282]:
horse_short_data.rectal_temp.isna().sum()

60

Количество пропусков - **60**

Метрики до заполнения пропущенных значений

In [283]:
get_metrics(horse_short_data_original.rectal_temp)

{'min': 35.4,
 'max': 40.8,
 'range': 5.4,
 'mean': 38.17,
 'median': 38.2,
 'std': 0.73,
 'var': 0.54,
 'mode': [38.0],
 'q1': 37.8,
 'q3': 38.5,
 'iqr': 0.7}

Попробуем заполнить пропущенные значения медианой

In [284]:
rectal_temp_with_fillna = horse_short_data.rectal_temp.fillna(horse_short_data.rectal_temp.median())
get_metrics(rectal_temp_with_fillna)

{'min': 35.4,
 'max': 40.8,
 'range': 5.4,
 'mean': 38.17,
 'median': 38.2,
 'std': 0.66,
 'var': 0.43,
 'mode': [38.2],
 'q1': 37.9,
 'q3': 38.5,
 'iqr': 0.6}

Метрики изменились совсем незначительно. Используем этот метод заполнения пропусков

In [ ]:
horse_short_data.rectal_temp = horse_short_data.rectal_temp.fillna(horse_short_data.rectal_temp.median())

Проверим на наличие пропущенных значений

In [287]:
horse_short_data.rectal_temp.isna().sum()

0

Пропусков нет

---

#### Пропуски в столбце Pulse

In [288]:
horse_short_data.pulse.isna().sum()

24

Количество пропусков - **24**

Метрики до заполнения пропущенных значений

In [289]:
get_metrics(horse_short_data.pulse)

{'min': 30.0,
 'max': 184.0,
 'range': 154.0,
 'mean': 71.93,
 'median': 64.0,
 'std': 28.68,
 'var': 822.57,
 'mode': [48.0],
 'q1': 48.0,
 'q3': 88.0,
 'iqr': 40.0}

Попробуем заполнить пропущенные значения медианой

In [290]:
pulse_with_fillna = horse_short_data.pulse.fillna(horse_short_data.pulse.median())
get_metrics(pulse_with_fillna)

{'min': 30.0,
 'max': 184.0,
 'range': 154.0,
 'mean': 71.3,
 'median': 64.0,
 'std': 27.59,
 'var': 760.99,
 'mode': [64.0],
 'q1': 48.0,
 'q3': 88.0,
 'iqr': 40.0}

Метрики изменились незначительно. Используем этот метод заполнения пропусков

In [ ]:
horse_short_data.pulse = horse_short_data.pulse.fillna(horse_short_data.pulse.median())

Проверим на наличие пропущенных значений

In [292]:
horse_short_data.pulse.isna().sum()

0

Пропусков нет

---

#### Пропуски в столбце Respiratory rate

In [293]:
horse_short_data.respiratory_rate.isna().sum()

58

Количество пропусков - **58**

Метрики до заполнения пропущенных значений

In [294]:
get_metrics(horse_short_data.respiratory_rate)

{'min': 8.0,
 'max': 96.0,
 'range': 88.0,
 'mean': 30.43,
 'median': 24.0,
 'std': 17.68,
 'var': 312.52,
 'mode': [20.0],
 'q1': 18.0,
 'q3': 36.0,
 'iqr': 18.0}

Попробуем заполнить пропущенные значения медианой

In [295]:
respiratory_rate_with_fillna = horse_short_data.respiratory_rate.fillna(horse_short_data.respiratory_rate.median())
get_metrics(respiratory_rate_with_fillna)

{'min': 8.0,
 'max': 96.0,
 'range': 88.0,
 'mean': 29.18,
 'median': 24.0,
 'std': 16.07,
 'var': 258.18,
 'mode': [24.0],
 'q1': 20.0,
 'q3': 34.5,
 'iqr': 14.5}

Метрики изменились незначительно. Используем этот метод заполнения пропусков

In [ ]:
horse_short_data.respiratory_rate = horse_short_data.respiratory_rate.fillna(horse_short_data.respiratory_rate.median())

Проверим на наличие пропущенных значений

In [298]:
horse_short_data.respiratory_rate.isna().sum()

0

Пропусков нет

---

#### Пропуски в столбце Total protein

In [299]:
horse_short_data.total_protein.isna().sum()

33

Количество пропусков - **33**

Метрики до заполнения пропущенных значений

In [300]:
get_metrics(horse_short_data.total_protein)

{'min': 3.3,
 'max': 89.0,
 'range': 85.7,
 'mean': 24.52,
 'median': 7.5,
 'std': 27.51,
 'var': 756.75,
 'mode': [6.5, 7.0],
 'q1': 6.5,
 'q3': 57.0,
 'iqr': 50.5}

Попробуем заполнить пропущенные значения медианой

In [301]:
total_protein_with_fillna = horse_short_data.total_protein.fillna(horse_short_data.total_protein.median())
get_metrics(total_protein_with_fillna)

{'min': 3.3,
 'max': 89.0,
 'range': 85.7,
 'mean': 22.64,
 'median': 7.5,
 'std': 26.49,
 'var': 701.48,
 'mode': [7.5],
 'q1': 6.6,
 'q3': 53.5,
 'iqr': 46.9}

Метрики изменились незначительно. Используем этот метод заполнения пропусков

In [ ]:
horse_short_data.total_protein = horse_short_data.total_protein.fillna(horse_short_data.total_protein.median())

Проверим на наличие пропущенных значений

In [303]:
horse_short_data.total_protein.isna().sum()

0

Пропусков нет

---

#### Пропуски в столбце Nasogastric reflux PH

In [304]:
horse_short_data.nasogastric_reflux_ph.isna().sum()

246

Количество пропусков - **246**

Пропусков очень много, более 80%. Можно попробовать заполнить столбец значениями на основе существующих данных, можно просто его удалить. Сложно сказать насколько полезны эти данные, но предположим, что не очень, так что удалим его

In [ ]:
horse_short_data.drop(['nasogastric_reflux_ph'], axis = 1, inplace=True)

---

#### Пропуски в столбце Pain

In [309]:
horse_short_data.pain.isna().sum()

55

Количество пропусков - **55**

Метрики до заполнения пропусков

In [311]:
get_metrics(horse_short_data.pain)

{'mode': [3.0]}

Распределение значений

In [310]:
horse_short_data.pain.value_counts()

3.0    67
2.0    59
5.0    41
4.0    39
1.0    38
Name: pain, dtype: int64

Мы видим распределение значений в выбранном столбце. Посколько нет явно выраженной моды, заменим пропущенные значения, основываясь на частотном распределении значений в столбце

In [313]:
pain_distribution = horse_short_data_original.pain.value_counts(normalize=True)
pain_distribution

3.0    0.274590
2.0    0.241803
5.0    0.168033
4.0    0.159836
1.0    0.155738
Name: pain, dtype: float64

In [314]:
pain_missing = horse_short_data_original.pain.isnull()

In [ ]:
horse_short_data.loc[pain_missing, 'pain'] = np.random.choice(pain_distribution.index, size=len(horse_short_data[pain_missing]),p=pain_distribution.values)

Проверим на наличие пропущенных значений

In [316]:
horse_short_data.pain.isna().sum()

0

Проверим не изменилась ли мода

In [318]:
get_metrics(horse_short_data.pain)

{'mode': [3.0]}

Мода осталась прежней

---

#### Пропуски в столбце Peripheral pulse

In [319]:
horse_short_data.peripheral_pulse.isna().sum()

69

Количество пропусков - **69**

Метрики до заполнения пропусков

In [320]:
get_metrics(horse_short_data.peripheral_pulse)

{'mode': [1.0]}

Распределение значений

In [72]:
horse_short_data.peripheral_pulse.value_counts()

1.0    115
3.0    102
4.0      8
2.0      5
Name: peripheral_pulse, dtype: int64

Мы видим распределение значений в выбранном столбце. Посколько нет явно выраженной моды, заменим пропущенные значения, основываясь на частотном распределении значений в столбце

In [321]:
peripheral_pulse_distribution = horse_short_data_original.peripheral_pulse.value_counts(normalize=True)
peripheral_pulse_distribution

1.0    0.500000
3.0    0.443478
4.0    0.034783
2.0    0.021739
Name: peripheral_pulse, dtype: float64

In [322]:
peripheral_pulse_missing = horse_short_data_original.peripheral_pulse.isnull()

In [323]:
horse_short_data.loc[peripheral_pulse_missing, 'peripheral_pulse'] = np.random.choice(peripheral_pulse_distribution.index, size=len(horse_short_data[peripheral_pulse_missing]),p=peripheral_pulse_distribution.values)

Проверим на наличие пропущенных значений

In [324]:
horse_short_data.peripheral_pulse.isna().sum()

0

Проверим не изменилась ли мода

In [325]:
get_metrics(horse_short_data.peripheral_pulse)

{'mode': [1.0]}

Мода осталась прежней

---

#### Пропуски в столбце Mucous membranes

In [148]:
horse_short_data_original.mucous_membranes.isna().sum()

46

Количество пропусков - **46**

Метрики до заполнения пропусков

In [326]:
get_metrics(horse_short_data_original.mucous_membranes)

{'mode': [1.0]}

Распределение значений

In [327]:
horse_short_data_original.mucous_membranes.value_counts()

1.0    79
3.0    58
4.0    41
2.0    30
5.0    25
6.0    20
Name: mucous_membranes, dtype: int64

Мы видим распределение значений в выбранном столбце. Посколько нет явно выраженной моды, заменим пропущенные значения, основываясь на частотном распределении значений в столбце

In [329]:
mucous_membranes_distribution = horse_short_data_original.mucous_membranes.value_counts(normalize=True)
mucous_membranes_distribution

1.0    0.312253
3.0    0.229249
4.0    0.162055
2.0    0.118577
5.0    0.098814
6.0    0.079051
Name: mucous_membranes, dtype: float64

In [330]:
mucous_membranes_missing = horse_short_data_original.mucous_membranes.isnull()

In [331]:
horse_short_data.loc[mucous_membranes_missing, 'mucous_membranes'] = np.random.choice(mucous_membranes_distribution.index, size=len(horse_short_data[mucous_membranes_missing]),p=mucous_membranes_distribution.values)

Проверим на наличие пропущенных значений

In [332]:
horse_short_data.mucous_membranes.isna().sum()

0

Проверим, не изменилась ли мода

In [333]:
get_metrics(horse_short_data.mucous_membranes)

{'mode': [1.0]}

Мода осталась прежней

---

#### Результирующий датафрейм с заполненными пропущенными значениями

In [334]:
horse_short_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   rectal_temp       299 non-null    float64 
 1   pulse             299 non-null    float64 
 2   respiratory_rate  299 non-null    float64 
 3   total_protein     299 non-null    float64 
 4   pain              299 non-null    category
 5   peripheral_pulse  299 non-null    category
 6   mucous_membranes  299 non-null    category
dtypes: category(3), float64(4)
memory usage: 11.0 KB


In [335]:
horse_short_data.head()

rectal_temp  pulse  respiratory_rate  total_protein pain peripheral_pulse  \
0         39.2   88.0              20.0           85.0  3.0              1.0   
1         38.3   40.0              24.0            6.7  3.0              1.0   
2         39.1  164.0              84.0            7.2  2.0              1.0   
3         37.3  104.0              35.0            7.4  3.0              1.0   
4         38.2   64.0              24.0            7.5  2.0              1.0   

  mucous_membranes  
0              4.0  
1              3.0  
2              6.0  
3              6.0  
4              3.0